queue/src/main/java/org/killbill/queue/DefaultReaper.java

```diff
@@ -0,0 +1,118 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.queue;
+
+import java.util.Date;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.killbill.clock.Clock;
+import org.killbill.commons.concurrent.Executors;
+import org.killbill.queue.api.PersistentQueueConfig;
+import org.killbill.queue.api.Reaper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class DefaultReaper implements Reaper {
+
+    private final DBBackedQueue<?> dao;
+    private final PersistentQueueConfig config;
+    private final ScheduledExecutorService scheduler;
+    private final Clock clock;
+    private final AtomicBoolean isStarted;
+    private final String threadScheduledExecutorName;
+    private ScheduledFuture<?> reapEntriesHandle;
+
+    private final long FIVE_MINUTES = 300000;
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultReaper.class);
+
+    public DefaultReaper(final DBBackedQueue<?> dao, final PersistentQueueConfig config, final Clock clock, final String threadScheduledExecutorName) {
+        this.dao = dao;
+        this.config = config;
+        this.clock = clock;
+        this.isStarted = new AtomicBoolean(false);
+        this.threadScheduledExecutorName = threadScheduledExecutorName;
+        this.scheduler = Executors.newSingleThreadScheduledExecutor(threadScheduledExecutorName);
+    }
+
+    @Override
+    public void start() {
+        if (!isStarted.compareAndSet(false, true)) {
+            return;
+        }
+
+        final long pendingPeriod = getReapThreshold();
+        final Runnable reapEntries = new Runnable() {
+            @Override
+            public void run() {
+                dao.reapEntries(getReapingDate());
+            }
+
+            private Date getReapingDate() {
+                return clock.getUTCNow().minusMillis((int) pendingPeriod).toDate();
+            }
+        };
+
+        reapEntriesHandle = scheduler.scheduleWithFixedDelay(reapEntries, pendingPeriod, pendingPeriod, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void stop() {
+        if (!isStarted.compareAndSet(true, false)) {
+            return;
+        }
+
+        if (!reapEntriesHandle.isCancelled() || !reapEntriesHandle.isDone()) {
+            reapEntriesHandle.cancel(true);
+        }
+
+        if (!scheduler.isShutdown()) {
+            scheduler.shutdown();
+            try {
+                scheduler.awaitTermination(5, TimeUnit.SECONDS);
+            } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+                log.info(String.format("%s stop sequence has been interrupted",threadScheduledExecutorName));
```

Comment:
You need to log first, then interrupt the thread.

Also, with SLF4J, you actually don't need `String.format`, you can do:

```
log.info("{} stop sequence has been interrupted",threadScheduledExecutorName);
```

You need to use `{}` and not `%s` though.